File: leetcode/28-find-the-index-of-the-first-occurrence-in-a-string.js
// https://leetcode.com/problems/find-the-index-of-the-first-occurrence-in-a-string/ (top-interview-questions)


/**
 * @param {string} haystack
 * @param {string} needle
 * @return {number}
 */
var strStr = function(haystack, needle) {

  // TODO: implement Knuth-Morris-Pratt KMP algorithm => O(n+m) time complexity solution

  // return haystack.indexOf(needle);

  for (let i = 0; i < haystack.length; i++) {
    if (haystack[i] === needle[0]) {
      let j = 1;
      while (j < needle.length && haystack[i+j] === needle[j]) j += 1;
      if (j === needle.length) return i;
    }
  }

  return -1;
};


strStr("sadbutsad", "sad");

File: leetcode/28-find-the-index-of-the-first-occurrence-in-a-string.py
# https://leetcode.com/problems/find-the-index-of-the-first-occurrence-in-a-string/ (top-interview-questions)

class Solution:
	def strStr(self, haystack: str, needle: str) -> int: # 
		if len(needle) == 0:
			return 0
		
		for i in range(len(haystack)-len(needle)+1):
			if haystack[i] == needle[0]:
				j = 1
				while j < len(needle) and haystack[i+j] == needle[j]:
					j += 1
				if j == len(needle):
					return i

		return -1
